### ISSUE TITLE
Feature Request: Allow Measures to be Invalid Duration While Editing

### ISSUE THREAD

**Issue by j-t-m-e** [helge17/tuxguitar#579]
Request for flexible bar duration, or to allow a bar to be too long or too short while editing.

As soon as I enter a note it automatically fills the rest of the bar with rests. This makes it difficult to edit the duration of notes after the fact because I find myself struggling with the strict limitation of not letting the bar be anything but exactly full.

Forcing all measures to be permanently valid greatly simplifies the job from software point of view, but allowing measures to be invalid duration would make editing and composing easier because you can under-fill or over-fill a measure temporarily while adjusting the duration of any note.

**Comment by belfie13** [comment]
yes this is a great feature i think. constantly running into problems where shifting one note will shift the entire score.

i'm not sure how much is involved with playing measures out of length but i'd be happy even if the measure is marked as incorrect length and it won't play.

**Comment by guiv42** [comment]
Yes, this is for sure an interesting option I'll investigate. But it's really complex, so it will take time.

@j-t-m-e just a question: I clearly understand the added value of this tolerance, but what I don't understand is why filling a measure with rests is a problem. Basically, if I want to create new notes, I need first to place the caret on an existing beat (e.g. a rest), then only I can add my note(s). For example, if at some point in time I have a 4/4 measure with one single quarter in it (temporarily invalid measure), how could I add new beats?

**Comment by guiv42** [comment]
In fact, it *is* currently possible to define invalid measures, but it's a *bug* (e.g. see #118). Note that it doesn't prevent song to be played or saved.
Before going any further on this feature request, an unambiguous evaluation of measure validity is definitely required: #118 needs to be fixed.

**Comment by belfie13** [comment]
> For example, if at some point in time I have a 4/4 measure with one single quarter in it (temporarily invalid measure), how could I add new beats?

either insert new beat or when navigating right over a bar line of a measure with less beats than the time signature it will automatically add a rest.

so if you have one single quarter in a 4/4 and your cursor is on the note, when you press right to go to the next measure it will add a rest instead.

and could add an action to fill the invalid measure with rests.

**Comment by guiv42** [comment]
I've started the development of this feature.
Basically the idea is to add an optional "free edition mode", disabled by default so that users are not confused. In this mode, some actions would be performed without any control. Identified so far:
- insert beat
- delete beat
- change note duration
- change measure time signature

I'm probably missing others. Please complete this list if you can think of any other action.

This raises an important question about user interface: when something is invalid in the song, it is of utmost importance to inform user *where* something is wrong, and precisely *what* is wrong. Else it could be *really* difficult for user to fix the problems.
Example: "track 2, measure 7: voice 2 is too long"

The question is: *where* in the user interface to display this information?
Keep in mind that many such errors can be present simultaneously, so some significant space is needed.

I'm currently thinking of creating a new window to display all errors, only displayed in free edition mode. It would contain the following content:
- boolean global status: song OK/KO
- a list of all errors. For each error:
  - a button "Go", moving the cursor to the invalid track/measure/voice
  - a button "Fix", enabled only if cursor is in the invalid measure, where TuxGuitar could try to fix automatically the issue

This is not so trivial, as for example errors could be present in masked tracks.

I would appreciate some feedback or suggestions

**Comment by guiv42** [comment]
I've started the development of this feature.
Basically the idea is to add an optional "free edition mode", disabled by default so that users are not confused. In this mode, some actions would be performed without any control. Identified so far:
- insert beat
- delete beat
- change note duration
- change measure time signature

I'm probably missing others. Please complete this list if you can think of any other action.

This raises an important question about user interface: when something is invalid in the song, it is of utmost importance to inform user *where* something is wrong, and precisely *what* is wrong. Else it could be *really* difficult for user to fix the problems.
Example: "track 2, measure 7: voice 2 is too long"

The question is: *where* in the user interface to display this information?
Keep in mind that many such errors can be present simultaneously, so some significant space is needed.

I'm currently thinking of creating a new window to display all errors, only displayed in free edition mode. It would contain the following content:
- boolean global status: song OK/KO
- a list of all errors. For each error:
  - a button "Go", moving the cursor to the invalid track/measure/voice
  - a button "Fix", enabled only if cursor is in the invalid measure, where TuxGuitar could try to fix automatically the issue

This is not so trivial, as for example errors could be present in masked tracks.

@j-t-m-e , @belfie13 , I would appreciate some feedback or suggestions

**Comment by guiv42** [comment]
> maybe consider putting the information in a toolbar? Or putting it in the editing toolbar?

Yes, I thought of a toolbar option also. It's a bit more complex from implementation point of view, as it would change the layout of the main window, but it could be more comfortable for user.
But: what kind of "shape" could such a toolbar have?
- it's not possible to use an horizontal toolbar (as the current main toolbar), as it would not be high enough to display everything
- I have the same problem with the current edit toolbar, which would not be wide enough.

Imagine there are 10 or more errors to display: probably a slider would become necessary, which is definitely not obvious in a toolbar.

> The "fix" feature sounds interesting!

I think it should be possible to do this in many cases, typically when measure is too long and holds many rests. I cannot promise it will work in all cases.

**Comment by guiv42** [comment]
Here is a basic mock-up to illustrate the concept.
@j-t-m-e , @belfie13 , please give me some feedback before I get any further (development of UIs is not my cup of tea, I would like to avoid future reworks).

Basically the idea is to add 2 icons in edit toolbar (and menu entries also). They are present in the screenshot below, but without any icon yet.
- activate/de-activate free edition mode
- display errors: this opens a dedicated dialog

@helge17, that would imply 4 more icons: 1 for each new button in edit toolbar, 1 for "OK" and 1 for "KO" statuses (the cross in the screenshot). Does it seem feasible?

![Image](https://github.com/user-attachments/assets/2aac12a3-19d5-4746-af4b-502e49bc2f93)

**Comment by helge17** [comment]
> @helge17, that would imply 4 more icons: 1 for each new button in edit toolbar, 1 for "OK" and 1 for "KO" statuses (the cross in the screenshot). Does it seem feasible?

Sure, I will take care of the icons.

**Comment by helge17** [comment]
I would suggest the following icons (Breeze, 44x44px & 22x22px):

**edit_mode_free.png**
![Image](https://github.com/user-attachments/assets/aaa315d0-ba37-4de9-87c7-64465353fd5e)
![Image](https://github.com/user-attachments/assets/8a6c7c8d-c52d-49cd-afbb-181ecf9514c0)
or
![Image](https://github.com/user-attachments/assets/50f09394-2d80-4e45-a016-c8ea485c83ce)
![Image](https://github.com/user-attachments/assets/9b37af50-5b6d-4b85-ae89-d11d55757cc5)

**measure_status_check.png**
![Image](https://github.com/user-attachments/assets/d0a474b5-c6cb-4753-bc4f-98d30f06e818)
![Image](https://github.com/user-attachments/assets/84df8cfd-c88d-484f-b750-c5ce7b4f42ce)
or
![Image](https://github.com/user-attachments/assets/c51e2fd7-926f-43fe-97c1-001f3783b27c)
![Image](https://github.com/user-attachments/assets/7b99f552-d312-4dab-837e-256ca5befa75)

I tend to use the tilde (≈ means “approximately”). What do you think?

**measure_status_ok.png**
![Image](https://github.com/user-attachments/assets/4a0268b8-d465-4cc1-823c-b42187e95c5b)
![Image](https://github.com/user-attachments/assets/2f81e9bb-a0d4-4416-add5-3028c5874641)

**measure_status_error.png**
![Image](https://github.com/user-attachments/assets/9cf92a00-d8e1-4527-a149-deaf432b7024)
![Image](https://github.com/user-attachments/assets/16eea936-b3d1-4466-a6fd-dff2ab27d3ca)

**Comment by guiv42** [comment]
Yes, that looks pretty good!

+1 for the tilde (or wavy lines).
For skins with colors (e.g. Oxygen) that could be interesting to show red lines in edit_mode_free.png, and red/green ok/ko symbols in measure_status_check.png?

**Comment by guiv42** [comment]
Development is progressing, I should be able to share a version for test in the coming days.

@helge17 : how do you think we should proceed?
It's definitely not mature enough to be merged, not even into tuxguitar-next branch. Modification is quite deep, it definitely deserves some testing. So I can propose the following options (other suggestions welcome!):

1. I publish a new branch in my repo. Drawback: I can only provide binaries for Linux/Windows, this will prevent beta-testing by macOS users
2. We create a new branch in this repo
3. We wait for tuxguitar-next to be merged into master... and then we re-create a new tuxguitar-next branch for this feature

**Comment by helge17** [comment]
I would vote for option 3. Tomorrow I will merge tuxguitar-next into master and then this new feature can be merged into tuxguitar-next.

**Comment by helge17** [comment]
tuxguitar-next is now merged into master.

**Comment by guiv42** [comment]
I just merged (#662) a very first version of this feature in tuxguitar-next branch. Work is in progress, it's not mature yet.

@j-t-m-e, @belfie13 (and anyone else!): help would be greatly appreciated for beta-testing.

**Limitations:**
- I did not perform any test outside of Linux/SWT configuration
- I did not check at all how this mode interferes with the player, I wouldn't be surprised by weird results. For simplicity we could eventually forbid to play an invalid song.
- I did not check either how this mode behaves when a beat range is selected by click & drag
- Multi-voices measures should be handled properly, but I did not perform so many tests in this configuration

**Still TODO:**
- new menu entries need to be created
- user documentation deserves a serious update, with appropriate user warnings like: "Don't use this mode if you don't have a correct knowledge of music theory. It will let you do anything, including errors that may be difficult to fix". To be done only after the feature is stabilized.

**Open points:**
- should we display a warning when saving an invalid song?
- should we simply forbid to save an invalid song? (see next point, that could be simpler)
- **what about Android app?** I'm not skilled enough in Android to develop such a user interface. And there is a problem if Android app opens files with invalid measures created by the desktop app: I did not test how song edition functions work when caret is in an invalid measure and free edition mode is disabled (this configuration cannot happen in desktop app)

**Comment by guiv42** [comment]
I've just detected another issue: copy/paste measure sometimes fails with an exception :(

**Comment by guiv42** [comment]
Issues understood, fix coming soon:
- copy/paste measure
- delete rest beat

However, I cannot reproduce the issue with left/right arrows.
@j-t-m-e : could you please describe a detailed procedure to reproduce the issue ?
I tried to create 4/4 measures with 14 quarters in them (!), and left/right arrows seem to work correctly:

https://github.com/user-attachments/assets/e9f9b6bc-142a-4032-a20f-4db86fdc9820

There must be another condition to trigger the issue.

**Comment by guiv42** [comment]
Thanks, very clear. Must be linked to the combination of [free edition mode] + [click & drag select], I've not really thought of these cases yet.
I'll debug that.

**Comment by guiv42** [comment]
That looks better now

https://github.com/user-attachments/assets/f1b0cbe6-cf74-4342-8bcd-5cecdcfe0882

**Comment by guiv42** [comment]
I had a try, but could not make Android app work handle invalid measures correctly. Misbehavior includes app crash.

Considering the behavior of old TG versions and external software (gp, musicxml, ...) is undefined, the conclusion is finally simple: TuxGuitar **shall not save/export/print invalid songs**.
Implemented (#671)

**Comment by belfie13** [comment]
there's 6 commits since tuxguitar-next release... how do i build the macos-swt version for the latest commit locally?

**Comment by guiv42** [comment]
You can as well test the last pre-release of tuxguitar-next. A few bugs were found already, all are listed in this discussion and now fixed.
But I would not be surprised if there are others.
Or you can wait for next pre-release.

**Comment by guiv42** [comment]
thank you @helge17  for the pre-release.
@belfie13  you can test today's pre-release it's fully up to date

**Comment by belfie13** [comment]
i'll be as often as i can but probably two days before i get a break and can help

**Comment by guiv42** [comment]
"remove measure" issue should be fixed now (#673)

**Comment by guiv42** [comment]
Thanks to this implementation, I just discovered that some of guitar pro tabs I found on the Internet included invalid measures!
Measure errors dialog automatically opens when user opens an invalid *tuxguitar song*, but not when user imports an invalid *foreign format song*, I'll try to fix that.

**Comment by guiv42** [comment]
In fact, it seems many gp files imported by TuxGuitar are invalid. I have several occurrences where measures are too long by a 32nd. It looks very much like these notes should have been grace notes.
I don't have a GP license, so I don't know if the original files were incorrect, or if it's just a bug of TuxGuitar when importing guitar pro files.
Request for help: could someone with a GuitarPro license try to open [this tab](https://tabs.ultimate-guitar.com/tab/dire-straits/where-do-you-think-youre-going-guitar-pro-1021680), and tell me if track4, measure 58 is valid or not? After import by TG it is invalid (one 32nd too long):

![Image](https://github.com/user-attachments/assets/dbc15c1e-3ffb-43ef-9654-e7e1377efdd5)

**Comment by guiv42** [comment]
Thank you @j-t-m-e . I was just looking at the code, there's just nothing about grace notes when TuxGuitar imports a .gp file. I'll create a dedicated issue for that.

It's strange, track 1 measures 42 & 45 are shown valid after TG import

![Image](https://github.com/user-attachments/assets/5c97a2cf-38bc-4aa7-b2a5-231ae97e125f)

Could you please show me what these measures look like in GP?

**Comment by guiv42** [comment]
OK, thanks, it's understandable: these measures are too short, and TuxGuitar automatically fills too short measures with rest (historical behavior, not linked to this issue).
See #680 for the grace notes issue.

**Comment by guiv42** [comment]
Good news, I could fix the edition issues with Android (#681)

**Comment by guiv42** [comment]
I think it could be pretty easy to add "Fix measure" action to Android app's "Measure" menu (see a similar modification done in 20adf3374bec4a286df729cfe35de6c64883e39e). Without the complete "measure errors" dialog (too complex for me)

@helge17 : can you please remind me how the translation strings are updated in Android app? Shall I add a string for this new menu item in translation xml files, or is it done automatically by the [perl script](https://github.com/helge17/tuxguitar/blob/master/misc/messages_desktop_to_android.pl)?

**Comment by guiv42** [comment]
@j-t-m-e , @belfie13 , could you please give me some feedback of latest pre-release ? (>=2025-04-06-tuxguitar-next)

Since there were some regressions (hopefully now fixed), I don't know if this feature is mature enough to be merged.

**Comment by guiv42** [comment]
OK, thanks. So I will spend some time to update user doc, then I will merge it (unless a new regression pops up)

**Comment by guiv42** [comment]
merged into master branch
